components/homme/src/share/prim_driver_base.F90

```diff
@@ -1468,113 +1426,114 @@ subroutine applyCAMforcing_tracers(elem,hvcoord,np1,np1_qdp,dt,adjustment)
+                     ps(i,j)=ps(i,j) + fq
+                     dp_adj(i,j,k)=dp_adj(i,j,k) + fq   !  ps =  ps0+sum(dp(k))
                   endif
                enddo
             end do
          end do
       end do
-
-      do k=1,nlev
-         do ic=1,qsize
+   else ! end of adjustment
+      ! apply forcing to Qdp
+      elem%derived%FQps(:,:)=0
+      do q=1,qsize
+         do k=1,nlev
             do j=1,np
                do i=1,np
-                  ! make Q consistent now that we have updated ps_v above
-                  ! recompute dp, since ps_v was changed above
-                  dp(i,j,k) = ( hvcoord%hyai(k+1) - hvcoord%hyai(k) )*hvcoord%ps0 + &
-                       ( hvcoord%hybi(k+1) - hvcoord%hybi(k) )*elem%state%ps_v(i,j,np1)
-                  elem%state%Q(i,j,k,ic)= &
-                       elem%state%Qdp(i,j,k,ic,np1_qdp)/dp(i,j,k)
-               end do
-            end do
-          end do
-        end do
-
-     else ! end of adjustment
-        ! apply forcing to Qdp
-        elem%derived%FQps(:,:)=0
-        do q=1,qsize
-           do k=1,nlev
-              do j=1,np
-                 do i=1,np
-                    v1 = dt*elem%derived%FQ(i,j,k,q)
-                    if (elem%state%Qdp(i,j,k,q,np1_qdp) + v1 < 0 .and. v1<0) then
-                       if (elem%state%Qdp(i,j,k,q,np1_qdp) < 0 ) then
-                          v1=0  ! Q already negative, dont make it more so
-                       else
-                          v1 = -elem%state%Qdp(i,j,k,q,np1_qdp)
-                       endif
-                    endif
-                    elem%state%Qdp(i,j,k,q,np1_qdp) = elem%state%Qdp(i,j,k,q,np1_qdp)+v1
-                    if (q==1) then
-                       elem%derived%FQps(i,j)=elem%derived%FQps(i,j)+v1/dt
-                    endif
-                 enddo
-              enddo
-           enddo
-        enddo
-
-        if (use_moisture) then
-           ! to conserve dry mass in the precese of Q1 forcing:
-           elem%state%ps_v(:,:,np1) = elem%state%ps_v(:,:,np1) + &
-             dt*elem%derived%FQps(:,:)
-        endif
-
-        ! Qdp(np1) and ps_v(np1) were updated by forcing - update Q(np1)
-        do k=1,nlev
-           dp(:,:,k) = ( hvcoord%hyai(k+1) - hvcoord%hyai(k) )*hvcoord%ps0 + &
-                ( hvcoord%hybi(k+1) - hvcoord%hybi(k))*elem%state%ps_v(:,:,np1)
-        enddo
-        do q=1,qsize
-           do k=1,nlev
-              elem%state%Q(:,:,k,q) = elem%state%Qdp(:,:,k,q,np1_qdp)/dp(:,:,k)
-           enddo
-        enddo
-
-     endif ! if adjustment
+                  v1 = dt*elem%derived%FQ(i,j,k,q)
+                  if (elem%state%Qdp(i,j,k,q,np1_qdp) + v1 < 0 .and. v1<0) then
+                     if (elem%state%Qdp(i,j,k,q,np1_qdp) < 0 ) then
+                        v1=0  ! Q already negative, dont make it more so
+                     else
+                        v1 = -elem%state%Qdp(i,j,k,q,np1_qdp)
+                     endif
+                  endif
```

Comment:
Just a comment on the above code that always confused me.  If we are not happy with negative Qdp then why do we even allow v1=0 to be a valid case thus allowing negative Qdp to propagate through the code?  Is it so that an error checker will catch it later?

Follow up, how do we feel about a case where `Qdp+v1 < 0` but `v1>0`?  My guess is that when we consider even the possibility of negative Qdp we're thinking of very small neg. numbers so its kind of an edge case and we just need to deal with it in some way.  But if `v1<0` inspires us to not make Qdp even more negative, but if `v1>0` then we trigger the "else" case and just bump v1 up so that Qdp+v1=0 then there is an inconsistentcy in how we see the relationship between the tendency and the value.

Reply:
That's true, if phrased that way it seems inconsistent.  But it can also be phrased this way:

If Q<0, then we neglect all the forcing.
if Q>0 but Q+fq<0, then we only apply a % of the forcing  
if Q+fq>0    we apply the full forcing

So in this sense I believe it is consistent.  We never change the sign of the forcing, all we do is reduce it if it would drive Q negative (but we never reduce it below zero).

Reply:
No objections to changing this in the future, but for this PR we need to keep it this way to maintain BFB.

Reply:
Oh absolutely.  I think the PR is ready to go.  I just brought it up as something I've thought everytime I've looked at this chunk of code.